From e8934c5385764c42c170802b0efcea83e5adeecd Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Thu, 02 Jul 2020 23:28:41 +0300
Subject: [PATCH] [SW-14904] secure boot: public key utlity

This utility helps creating the needed device tree binary blob
containing the public key information needed by mkimage.

Usage:
create_pub_key_itb.sh /path/to/file.crt

The result will be stored in /path/to/file.dtb

Change-Id: I152c9a440d7e721450632cc79c1a1e5df5b14bec
---

diff --git a/scripts/create_pub_key_dtb.sh b/scripts/create_pub_key_dtb.sh
new file mode 100755
index 0000000..d8738d6
--- /dev/null
+++ b/scripts/create_pub_key_dtb.sh
@@ -0,0 +1,118 @@
+#! /bin/bash
+
+set -e
+
+if [ $# -ne 1 ]; then
+	echo "Usage $0 <crt file>" >&2
+	exit 1
+fi
+
+CRT_FILE=$1
+
+if [ ! -f ${CRT_FILE} ]; then
+	echo "File not exists ${CRT_FILE}" >&2
+	exit 1
+fi
+
+TMP_DIR=$(mktemp -d)
+CRT_FILE_NAME=$(basename ${CRT_FILE})
+CRT_NAME=${CRT_FILE_NAME%%.*}.crt
+KEY_NAME=${CRT_FILE_NAME%%.*}.key
+DTB_NAME=${CRT_FILE_NAME%%.*}.dtb
+OUTPUT_DIR=$(dirname "${BASH_SOURCE[0]}")/../output
+
+source ${OUTPUT_DIR}/.hl-vars.mk
+MKIMAGE=$(realpath ${OUTPUT_DIR}/${HL_DEV}/build/uboot-custom/tools/mkimage)
+DTC=$(realpath ${OUTPUT_DIR}/${HL_DEV}/build/uboot-custom/scripts/dtc/dtc)
+
+if [ ! -f ${MKIMAGE} ]; then
+	echo "Cannot find mkimage (${MKIMAGE})"
+	exit 1
+fi
+
+if [ ! -f ${DTC} ]; then
+	echo "Cannot find dtc (${DTC})"
+	exit 1
+fi
+
+echo "
+-----BEGIN PRIVATE KEY-----
+MIIG/wIBADANBgkqhkiG9w0BAQEFAASCBukwggblAgEAAoIBgQDv/CTbqZ0xfDO6
+dNH4gomexARqrNmYGJhR9ZUjizsfRPIW29t8nuhpwDMKEfVC7DYepIDAK46P03Va
+U1c38KMLpmiJR8jW2/4wEeSfCTqnplH3N5BL7twA8oa+Y+vlKF+QvQ5K3/IzrNDH
+U1ZXoArr8QbduvWS4sUhIpFeRXb+CIySjCxmK+wCtaL90P29Ebb/mNI8UzS55CN5
+GnwCdl4+W24k/khKWXaOO2Ab2M1WLz08BHs30W0386I1vQUqfIOIkKxk4rk40mQA
+0Owz8EjE6GfdM13G3yF3Ccd56zODsE+JbTtdy4ypkLmOp+kgUbHz5ls1Fx4WQH8P
+rwDp+xFb6/Mq1yjOM5oIkIoLmsnbl/n21XUpTT/Xt5uf8+taDf8atrz7uSDgTc5u
+CTTZA9P5vftcjLX4lLYCnc1mz5Qn5cKQEsJSxkifiaeGahGd+ZqhLhNwB9oh6Kfk
+rzePjacThuOFpcg1g06xMwLR/3IIaAhzIf8VK2ik1wdRDhgEzHMCAwEAAQKCAYEA
+mujn4eNa1NaXjYRbBU8+Bn4BW9q3mrZFnKIp4we84jJfEuJYi9kHrTujtpxg9RIE
+3ULS2TlEjRKybdveMP3kQWDpmbP8sLTRH7ZtHTuX4SA7SJlVS66j1Pi4ga1+WC33
+gjaHQ03Lyg+DtSbWVTm50xOCmx5T5rz7AqiNrYypgtGB7BcPkeRINHeK/swo/iKx
+Pcwn6UjMgDIDs+iko/JEY3whXT/G6jdVj/pDz7Xi1f8oR9WctCs7lSJvUsDw1LwJ
+68AdQH0ADbDiZfCXREYRtClKcch6uisdbKs/JUVwaWDB6r15Itzt16Dju9vEuINL
+2A8v9ILFI6lrev3YqaD2KnHRh7AOvVC7G+TetjempIJaJUSwnwp3izCp8xgXQjw5
+VlrzNzLnXITOxoyc+3FX1vzlR/K+sjduidOlM4i7IYWoozcnNci7/yHMNYWJJqOP
+TTuijFUHKDkhqbh0u/MVFbcMapgzyNIBaoxOwpp4AC4Fg/CPignm4FMYWi3vksQB
+AoHBAP66lkTqU6jlVLuBZbwP9ouXtD5ArIVWBUu46FcsWnzEXcBj/Mlo5xbjEBDM
+tBCOHOT4+FipgBZbEdiEDdzPNW1ZcJJ5qBpWTinK+/wGHHNzOHLNG4Y4CAdLt890
+n5DYSQ+qPOOFnHnqBqOZaTdqB/7PuLkskuhSjYr32nqd4An/GCmEAuZg2fyApnal
+yuTqn/qqonLbL+b9SyEToxP08PFaiavX8PwsavtSCFTlmn+vnC8GiyMFMOtrafUJ
+s0hxCQKBwQDxLrjIq7WqvYTekDhNe+ErZyxKyogv4OHudWYzXw4wpTIBlZ1FPocL
+z8WXCIRVbZO7G9AOFPc7QeXUi6dKZuwhEMqruGFx1Zir18pQgq/yi8iicSx+tQhu
+98SdXMOv6v2UairHzRZQ0kiwta4l98rElZfQJ/B96/slh3SfO32j5jeDzKbAiSYq
+aDWXmUt9iEMqg4odXuUde8MvgQ9TE8Ro+WjAaJLcWrNR39tqQu1yzU1eHaimeTB/
+XPEguTMofJsCgcBnVJdKg2fGdqkFmvsC69LOU47JRgD14awqSGrvmc1nGV0Sgzzz
+EvvTG+7TccaFhLCR7bEIflG4P98gXoklDmUU45g3wuP/Mw1oBLBtTqlR5+Z5I1zL
++DTBOOTqYdusrSAOG7exb+zGw7lHuh6lXPy7T9YWP8+kMIC8Y8ibvtk4FOPfS27F
+t3oV5LYiISUJ8WdODtPxFZr777n/Q/D+Unk+3sREEkfuc7I+GsvIbI9cxobryJK+
+dQegKCWGQhYZB8kCgcEAs+meflNY2V022K/M0EAw4Ce+qqwam1yuLSv2aSvncfJJ
+M1VgVF8obm83fkskLHXpVhKRSW/qXZHy2HK1AJN2YdQc26xXRIuJr+bxTM2wYsZT
+jJnQD1UMTXsWeCzS7U/miegIsbMTiobTIr7kgf4ujCbrcKPaipSWbJB3NrEyMCl0
+ECWB3u5wQgIquEPqRGJ5eCVXykEXuL6yA6XKwLZmNciXk7BunWIrkk/jZYk3iAQO
+Sr7lKAoPW3MBjN+aggExAoHBAK9gLe1oH3YGz1T3O9zDumXf9qr7RSC0SHk2k9Rt
+uypyKruT0AsJI7QOCXYfP27sOE1gwLe68XbETDoROCoYAWYYQuUPRGHOP8KXWbCw
+6kVg3x1V22gu6mmntxga03439tcsVpug86C2kSgoFXb1XeTuWGL2/ceAkLbZhL8N
+iViJIKUtGq4uFBxZfYoeUj0GPtHVqFB0bViGk7Vp4jpESforLL3jGZTRKPlbhXgB
+r8MbcYAhQxYUn2NiOTRHIsFM0g==
+-----END PRIVATE KEY-----
+" > ${TMP_DIR}/${KEY_NAME}
+
+cp ${CRT_FILE} ${TMP_DIR}/${CRT_NAME}
+
+pushd ${TMP_DIR} > /dev/null
+	echo "/dts-v1/; / {};" | ${DTC} -I dts -O dtb -o ${DTB_NAME} -
+	echo "
+	/dts-v1/;
+       	/ {
+		#address-cells = <2>;
+		description = \"dummy\";
+		images
+		{
+			dummy
+			{
+				data = /incbin/(\"${KEY_NAME}\");
+				type = \"standalone\";
+				compression = \"none\";
+				hash {
+					algo = \"sha256\";
+				};
+			};
+		};
+		configurations
+		{
+			conf
+			{
+				loadables = \"dummy\";
+				signature
+				{
+					algo = \"sha256,rsa3072\";
+					key-name-hint = \"${CRT_FILE_NAME%%.*}\";
+					sign-images = \"loadables\";
+				};
+			};
+		};
+	};" | ${MKIMAGE} -D "-W no-unit_address_vs_reg" -f - -r -k ${TMP_DIR} -K ${DTB_NAME} - > /dev/null
+popd > /dev/null
+
+cp -v ${TMP_DIR}/${DTB_NAME} $(dirname ${CRT_FILE})
